core/src/main/java/io/questdb/cairo/wal/WalTxnYieldEventsImpl.java

```diff
@@ -0,0 +1,133 @@
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.wal;
+
+import io.questdb.cairo.*;
+import io.questdb.network.YieldEvent;
+import io.questdb.network.YieldEventFactory;
+import io.questdb.std.FilesFacade;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjList;
+import io.questdb.std.datetime.millitime.MillisecondClock;
+import io.questdb.std.str.Path;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class WalTxnYieldEventsImpl implements WalTxnYieldEvents {
+
+    private final CharSequence dbRoot;
+    private final ConcurrentHashMap<TableToken, ObjList<YieldEvent>> events = new ConcurrentHashMap<>();
+    private final FilesFacade ff;
+    private final MillisecondClock millisecondClock;
+    private final long spinLockTimeout;
+    private final TxReader txReader;
+    private final YieldEventFactory yieldEventFactory;
+
+    public WalTxnYieldEventsImpl(@NotNull CairoConfiguration cairoConfiguration, @NotNull YieldEventFactory yieldEventFactory) {
+        this.dbRoot = cairoConfiguration.getRoot();
+        this.ff = cairoConfiguration.getFilesFacade();
+        this.txReader = new TxReader(ff);
+        this.millisecondClock = cairoConfiguration.getMillisecondClock();
+        this.spinLockTimeout = cairoConfiguration.getSpinLockTimeout();
+        this.yieldEventFactory = yieldEventFactory;
+    }
+
+    @Override
+    public void close() {
+        for (Map.Entry<TableToken, ObjList<YieldEvent>> entry : events.entrySet()) {
+            Misc.freeObjListAndClear(entry.getValue());
+        }
+    }
+
+    @Override
+    public @Nullable YieldEvent register(TableToken tableToken, long txn) {
+        Path tlPath = Path.PATH.get().of(dbRoot);
+        tlPath.trimTo(dbRoot.length()).concat(tableToken).concat(TableUtils.META_FILE_NAME).$();
+        if (ff.exists(tlPath)) {
+            long currentTxn = readSeqTxn(tableToken, tlPath);
+            if (currentTxn == -1 || currentTxn >= txn) {
+                // The txn is already visible or the table is dropped.
+                return null;
+            }
+
+            final YieldEvent yieldEvent = yieldEventFactory.newInstance();
+            events.compute(tableToken, (t, list) -> {
```

Comment:
implicit allocation

Reply:
Yes, but this one is essential since the lambda has to be a capturing one. If you have any suggestions on how to avoid that, I'm all for it.

Reply:
How about creating a static inner class and using an member or Thread local ? 

``
    static class MyBiFunc implements BiFunction<String, ObjList<String>, ObjList<String>> {
        String value;

        @Override
        public ObjList<String> apply(String s, ObjList<String> list) {
            if (list == null) {
                list = new ObjList<>();
            }
            list.add(value);
            return list;
        }

        public void setValue(String value) {
            this.value = value;
        }
    }
``
and then 

``
func.setValue(yieldToken);
events.compute(tableToken, func); 
``

Reply:
I also considered TL, but decided that it would make the code more complex while it already allocates event and exception objects. It's probably worth it to try, thanks for the suggestion. Will give it a go.

Reply:
Introduced TL in fa493333ca3bb0045b8f5cef9fd9d9fa4afb96ae

BTW the above `MyBiFunc` class won't do since it's not thread-safe: `WalTxnYieldEventsImpl` is called concurrently, so a single `func` instance won't be sufficient unless we use TL.

Reply:
Sorry, forgot to push the changes. Should be fixed now.

Reply:
@bziobrowski could you check the update? Is it ok if I resolve your comments?